File: matlab_scripts/PlotReferencePerformanceCycleFigure_Cell68.m
%% PlotReferencePerformanceCycleFigure_Cell68.m
% Summary: Reproduces the Reference Performance Cycle figure for Cell_68. The
%          driver loads and preprocesses the cell's cyclic-ageing CSV, selects
%          the April 2024 reference-cycle window, and delegates plotting to the
%          shared RPC figure function.
%
% Usage: Set DataRoot if needed, then run this script with no arguments. The
%        cell and reference-cycle window are fixed for this figure.
%
% Outputs: NextBMS_ReferencePerformanceCycle.png and
%          NextBMS_ReferencePerformanceCycle.pdf in this script's R-022 output
%          directory.
%
% Authors: GitHub Copilot.
% Dependency files: Functions/loadAndPreprocessAgeingCsv.m,
%                   Functions/getCellLabel.m, Functions/getFigureOutputDir.m,
%                   Functions/plotReferencePerformanceCycle.m.
% Last documented: 2026-09-01

clear; close all; clc;

%% Configuration
% Resolve the shared Functions folder relative to this script so it works from
% any current directory (same pattern as ExtractAgeingData.m).
scriptDir = fileparts(mfilename('fullpath'));
functionsDir = fullfile(scriptDir, '..', '..', 'Functions');
if ~exist(functionsDir, 'dir')
    functionsDir = fullfile(scriptDir, '..', 'Functions');
end
if exist(functionsDir, 'dir')
    addpath(functionsDir);
else
    error('Shared Functions folder not found from %s.', scriptDir);
end

% The RPC figure and every hard-coded zone/annotation timestamp inside
% plotReferencePerformanceCycle.m are specific to this cell and window - do not
% change the cell without also updating those timestamps in the plotting function.
cellNum       = 'Cell_68';
% DataRoot: single switch to the dataset root holding 1_Teardown/2_HalfCell/
% 3_Characterization/4_Ageing. Change this one line to retarget the script.
DataRoot      = '\\tsn.tno.nl\RA-Data\SV\sv-072952\BTS Data\NEXTBMS\ZenodoRoot';
DesiredFolder = fullfile(DataRoot, '4_Ageing', 'Cyclic_ageing_data');

% Publication zoom window for the Reference Performance Cycle view (identical to
% the window used in ExtractAgeingData.m for this cell).
referenceCycleStartTime = datetime(2024, 4, 23, 5, 51, 16);
referenceCycleEndTime   = datetime(2024, 4, 30, 4, 55, 12);

%% Load + preprocess the cell CSV via the shared ingestion helper
% loadAndPreprocessAgeingCsv reproduces the exact ingestion pipeline used by
% ExtractAgeingData.m (read + time-vector reconstruction + NaN-gap insertion +
% cumulative-charge integral), so the figure matches the one that script produces.
loadName  = fullfile(DesiredFolder, cellNum, [cellNum '.csv']);
cellLabel = getCellLabel(cellNum);
fprintf('Loading and preprocessing %s ...\n', cellNum);
[timeWithGaps, timeS, voltage, current, cellTemp, chamberTemp, cumulative_integral] = ...
    loadAndPreprocessAgeingCsv(loadName);  %#ok<ASGLU>  timeS is unused by the RPC plot

%% Produce the Reference Performance Cycle figure (and its vector PDF)
% plotReferencePerformanceCycle builds the 3-panel figure, the zone arrows, the
% voltage/current zoom insets and exports NextBMS_ReferencePerformanceCycle.pdf.
plotReferencePerformanceCycle(timeWithGaps, current, voltage, cellTemp, chamberTemp, ...
    cumulative_integral, cellNum, cellLabel, referenceCycleStartTime, referenceCycleEndTime, ...
    getFigureOutputDir('PlotReferencePerformanceCycleFigure_Cell68'));

fprintf('Reference Performance Cycle figure generated for %s.\n', cellNum);

File: python_scripts/PlotReferencePerformanceCycleFigure_Cell68.py
"""
PlotReferencePerformanceCycleFigure_Cell68 - Standalone driver that reproduces
the Reference Performance Cycle (RPC) publication figure for Cell_68 without
running the full ExtractAgeingData pipeline.

Python counterpart of matlab_scripts/PlotReferencePerformanceCycleFigure_Cell68.m
(todo #029).

Author: GitHub Copilot (for Robinson Medina / Feye Hoekstra)
Date: 2026-08-24

Compliance: R-004, R-013, R-016.
"""

import os
import sys
from pathlib import Path

import pandas as pd

# Resolve Functions/ regardless of caller's cwd or repository layout.
_SCRIPT_DIR = Path(__file__).resolve().parent
_FUNCTIONS_CANDIDATES = (_SCRIPT_DIR.parent.parent / 'Functions', _SCRIPT_DIR.parent / 'Functions')
for _FUNCTIONS_DIR in _FUNCTIONS_CANDIDATES:
    if (_FUNCTIONS_DIR / "get_figure_output_dir.py").exists():
        if str(_FUNCTIONS_DIR) not in sys.path:
            sys.path.append(str(_FUNCTIONS_DIR))
        break
else:
    raise FileNotFoundError(f"Shared Functions folder not found from {_SCRIPT_DIR}")

from loadAndPreprocessAgeingCsv import load_and_preprocess_ageing_csv  # noqa: E402
from plotReferencePerformanceCycle import plot_reference_performance_cycle  # noqa: E402
from getCellLabel import get_cell_label  # noqa: E402
from get_figure_output_dir import get_figure_output_dir  # noqa: E402


def main():
    """Load Cell_68's cyclic-ageing CSV and render the RPC publication figure."""

    # Fixed single-cell scope by design (R-025 plain Cell_<n> form).
    cell_num = 'Cell_68'

    # DataRoot: single switch to the dataset root (todo #051/#041 convention).
    data_root = r'\\tsn.tno.nl\RA-Data\SV\sv-072952\BTS Data\NEXTBMS\ZenodoRoot'
    desired_folder = os.path.join(data_root, '4_Ageing', 'Cyclic_ageing_data')
    csv_path = os.path.join(desired_folder, cell_num, f'{cell_num}.csv')

    cell_label = get_cell_label(cell_num)

    print(f'Loading and preprocessing {cell_num}...')
    (time_with_gaps, _time_s, voltage, current, cell_temp, chamber_temp,
     cumulative_integral) = load_and_preprocess_ageing_csv(csv_path)

    # Fixed publication zoom window for the Reference Performance Cycle view.
    reference_cycle_start_time = pd.Timestamp(2024, 4, 23, 5, 51, 16)
    reference_cycle_end_time = pd.Timestamp(2024, 4, 30, 4, 55, 12)

    fig = plot_reference_performance_cycle(
        time_with_gaps, current, voltage, cell_temp, chamber_temp,
        cumulative_integral, cell_num, cell_label,
        reference_cycle_start_time, reference_cycle_end_time,
        get_figure_output_dir('PlotReferencePerformanceCycleFigure_Cell68'),
    )
    fig.show()


if __name__ == '__main__':
    main()
